### PR TITLE
Add Razer BlackWidow Elite support

### PR DESCRIPTION
**Pull request by beranat**
[Razer BlackWidow Elite keyboard](https://www.razer.com/gaming-keyboards-keypads/razer-blackwidow-elite) support in driver and daemon.

Change log:

- Fully supported effects None, Static, Breath, Spectrum, Reactive and Starlight
- Partially supported effect Wave - 'Left-to-Right' only
- All multimedia button and spinner also works (generic (device_type=0:0) and driver (device_type=3:0) modes)
- Added scripts/driver/blackwidow_elite/test.py for checking
- Added writing possibility to 'test' device - it sends command direct to keyboard

[Continue for pull-request](https://github.com/openrazer/openrazer/pull/804), during some issues.
Sorry for inconvenience.

### CODE REVIEW

**Comment by magbest** [comment]
i am a bit of a noob. How do i add this patch to my polychromatic software?

**Comment by lah7** [comment]
If you're running Ubuntu/Debian, run this from the **Terminal**:

    git clone https://github.com/beranat/openrazer.git
    cd openrazer
    ./scripts/build_debs.sh
    dpkg -i ./dist/*.deb

See your other comment on https://github.com/polychromatic/polychromatic/pull/200 for patching Polychromatic.

**Comment by magbest** [comment]
@lah7 Thank you very much!!

**Comment by aelsenaar** [comment]
Hi, I installed your patched version of the driver. At first, it seems to work fine but after some time the issues the following error:

`May 24 11:49:16 seti openrazer-daemon[1707]: 2019-05-24 11:49:16 | razer.device0                  | DEBUG    | DBus call get_brightness
May 24 11:49:16 seti kernel: [ 9308.788780] razer driver: Device data transfer failed.
May 24 11:49:16 seti kernel: [ 9308.789284] razer driver: Invalid USB response. USB Report length: -32
May 24 11:49:16 seti kernel: [ 9308.789289] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
`

When I unplug the keyboard and plug it in again it works again for a short time. What can I do to help debug this issue?

What did I do:
* I clone-ed you repo
* build the driver
* created the deps packages and installed them
* rebooted the computer

I am using the Kubuntu distro with kernel version 5.0.0-15.

**Comment by beranat** [comment]
Dear @aelsenaar, Could you check - does the keyboard works under Windows well?

**Comment by aelsenaar** [comment]
@beranat, I checked the keyboard on a Windows laptop. Installed the default Razer software and updated the keyboard settings intensify. I switched the effects a lot of time and turned lighting on/off frequently. All the time I noticed no problems.

**Comment by beranat** [comment]
Dear **@aelsenaar**, thank for patience and cooperation, I will try to switch lights/brightness harder.
Does any one have the same issue?

**Comment by Rincewind70** [comment]
Tested it on Linux Mint 19.1 with Kernel 5.0.0-15. Works fine for me except for two things when trying to assign custom colors to keys (by the way, I am using a german keyopard layout):

- Assigning color to space bar does not work.
- Assigning color to Alt Gr will set the color of the razer logo instead. Alt Gr will stay dark.

When using a standard effect all keys and logo light up. Tested with RazerGenie and Polychromatic.

Regards
Rince

**Comment by beranat** [comment]
Dear **Rincewind70**, you should use a patched Polychromatic from [BlackWidow Elite support](https://github.com/polychromatic/polychromatic/pull/200).

**Comment by franco-ruggeri** [comment]
Hello @beranat, thank you for the support, I am testing (and enjoying) it...
On Ubuntu 18.04, with the ripple effect, after locking (super+L) and unlocking the lighting becomes off and the polychromatic daemon crashes.

File: driver/razerkbd_driver.c [hunk 1]
```diff
@@ -2444,11 +2589,14 @@ MODULE_DEVICE_TABLE(hid, razer_devices);
...
     .id_table = razer_devices,
+
+    .input_mapping = razer_kbd_input_mapping,
+
     .probe = razer_kbd_probe,
     .remove = razer_kbd_disconnect,
 
     .event = razer_event,
-    .raw_event = razer_raw_event,
+    .raw_event = razer_raw_event
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Please keep the comma

**Comment by z3ntu** [comment]
Should I review this? Or is it still WIP?

File: driver/razerkbd_driver.c (hunk 1 above)
Comment (beranat) [reply driver/razerkbd_driver.c]: Sorry, I will add

**Comment by beranat** [comment]
Dear *Luca Weiss (@z3ntu)*, tha main work is over, but I try to solve @aelsenaar's issue with any luck yet.
  
Dear **The Keyboard Owners**, could you check [aelsenaar's issue](https://github.com/openrazer/openrazer/pull/831/#issuecomment-495916259)? I can not reproduce this trouble. Has anyone the same problem?

**Comment by bmarshall1985** [comment]
@beranat I have been testing for a few days now and cannot recreate the issue @aelsenaar described. I am on 5.0.0.15 with pop os 19.04

**Comment by aelsenaar** [comment]
> Dear _Luca Weiss (@z3ntu)_, tha main work is over, but I try to solve @aelsenaar's issue with any luck yet.
> 
> Dear **The Keyboard Owners**, could you check [aelsenaar's issue](https://github.com/openrazer/openrazer/pull/831/#issuecomment-495916259)? I can not reproduce this trouble. Has anyone the same problem?

@beranat Last week I have installed Ubuntu 19.04 (not Kubuntu!) with kernel version 5.0.0-16-generic. I have not experienced the problem in the last 7 days. So I don't know what was going wrong with my previous setup but currently, everything seems to work fine.

**Comment by beranat** [comment]
Dear **Luca Weiss (@z3ntu)**, looks like development is over.

**Comment by DangerBlack** [comment]
Noob guide: https://github.com/openrazer/openrazer/pull/804#issuecomment-503786270

**Comment by DangerBlack** [comment]
Issue:
Ripple effect reset each reboot.

**Comment by tildearrow** [comment]
That's because the ripple effect is software-based.

**Comment by DangerBlack** [comment]
> That's because the ripple effect is software-based.

Is weird that after reboot the effect won't get reloaded properly.

**Comment by tildearrow** [comment]
> Is weird that after reboot the effect won't get reloaded properly.

The daemon still doesn't have the ability to remember effects and set them when the daemon starts, but I am working on it in a fork.

**Comment by laurinneff** [comment]
How do I build a pacman package?

File: driver/razerkbd_driver.c [hunk 2]
```diff
@@ -1708,12 +1821,6 @@ static int razer_raw_event(struct hid_device *hdev, struct hid_report *report, u
         return 0;
     }
 
-    // ...and on these keyboards
-    switch (usb_dev->descriptor.idProduct) {
-    case USB_DEVICE_ID_RAZER_HUNTSMAN_ELITE:
-        return 0;
-    }
-
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:1732]: Why did you remove this?

File: driver/razerkbd_driver.c [hunk 3]
```diff
@@ -1248,10 +1330,17 @@ static ssize_t razer_attr_write_set_fn_toggle(struct device *dev, struct device_
 /**
  * Write device file "test"
  *
- * Does nothing
+ * Send data direct to usb (but caclulate check sum)
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Typo in `caclulate`

File: driver/razerkbd_driver.c [hunk 4]
```diff
@@ -1734,22 +1841,33 @@ static int razer_raw_event(struct hid_device *hdev, struct hid_report *report, u
...
+                break;
+            case 0x53: // Next (song)
+                cur_value = USB_HID_KEY_MEDIA_NEXTSONG;
+                break;
+            case 0x55: // Play/Pause
+                cur_value = USB_HID_KEY_MEDIA_PLAYPAUSE;
+                break;
+            case 0x54: // Prev (song)
+                cur_value = USB_HID_KEY_MEDIA_PREVIOUSSONG;
                 break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:1882]: I hope these new cases don't break existing keyboards...

File: driver/razerkbd_driver.c [hunk 5]
```diff
@@ -1766,6 +1884,30 @@ static int razer_raw_event(struct hid_device *hdev, struct hid_report *report, u
     return 0;
 }
 
+/**
+ * Set static hid-events translation map
+ *
+ * These keyboards Generates wheel-events for volume control spinner
+ * - BlackWidow Elite
+ * - Huntsman Elite
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Please make the comment generic because the comment probably won't be updated with new devices

File: driver/razerkbd_driver.c [hunk 6]
```diff
@@ -2454,6 +2599,9 @@ MODULE_DEVICE_TABLE(hid, razer_devices);
 static struct hid_driver razer_kbd_driver = {
     .name = "razerkbd",
     .id_table = razer_devices,
+
+    .input_mapping = razer_kbd_input_mapping,
+
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Remove the added empty lines

File: driver/usb_hid_keys.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
+// USB HID keycodes
+// https://source.android.com/devices/input/keyboard-devices.html
```
Comment (z3ntu) [inline driver/usb_hid_keys.h]: I think it's better to reference e.g. http://www.freebsddiary.org/APC/usb_hid_usages.php

File: driver/usb_hid_keys.h [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+#pragma once
```
Comment (z3ntu) [inline driver/usb_hid_keys.h]: Remove the `#pragma once` as you have the `#ifndef` below as header guard

File: scripts/driver/blackwidow_elite/test.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/python3
```
Comment (z3ntu) [inline scripts/driver/blackwidow_elite/test.py:1]: Do we need this file in the tree?

**Review by z3ntu — CHANGES_REQUESTED** [review]
See comments, LGTM otherwise

**Comment by 123swk123** [comment]
support needed for JP layout, razergenie falls back to default layout

`cat /sys/bus/hid/drivers/razerkbd/0*/kbd_layout`
`0C`

File: driver/razerkbd_driver.c (hunk 2 above)
Comment (beranat) [reply driver/razerkbd_driver.c:1732]: Yes, this change has been made for support Multimedia buttons in Huntsman Elite (they are the same as Blackwidow Elite)

File: driver/razerkbd_driver.c (hunk 3 above)
Comment (beranat) [reply driver/razerkbd_driver.c]: thnx! Fixed

File: driver/razerkbd_driver.c (hunk 5 above)
Comment (beranat) [reply driver/razerkbd_driver.c]: +1, Thank you for your attention

File: driver/razerkbd_driver.c (hunk 6 above)
Comment (beranat) [reply driver/razerkbd_driver.c]: no problem, empty lines removed

File: driver/usb_hid_keys.h (hunk 7 above)
Comment (beranat) [reply driver/usb_hid_keys.h]: OK, I will replace link - I have used first founded one

File: driver/usb_hid_keys.h (hunk 8 above)
Comment (beranat) [reply driver/usb_hid_keys.h]: 'pragma once' increases compiling speed and works better IDE's code completion, but it may be not implemented in some compilers.

I have used both (pragma and ifndef-define) as a compromise between "works everywhere" and "works better".

File: scripts/driver/blackwidow_elite/test.py (hunk 9 above)
Comment (beranat) [reply scripts/driver/blackwidow_elite/test.py:1]: This script helps to check all keyboards features and has been written as others test scripts (for blackwidow_chroma_keyboard, ornata_chroma_keyboard and etc).

File: driver/razerkbd_driver.c (hunk 4 above)
Comment (beranat) [reply driver/razerkbd_driver.c:1882]: The code may effect only for *keyboards*, because original code with special scan codes 0x20..0x24 works for all keyboards (even boards without M1-M5 buttons) without any issues - I , suppose that handling special scan codes 0x52..0x54 for multimedia buttons should work for keyboards without multimedia buttons too

**Comment by beranat** [comment]
Dear **@z3ntu**, sorry for long delay - I fix code and wait for your comments

**Comment by z3ntu** [comment]
@beranat I've squashed the commits into one and fixed a few other formatting things. Could you please test that version to make sure I didn't break anything? After then it can be merged imo.

**Review by z3ntu — APPROVED** [review]
LGTM

**Comment by beranat** [comment]
Dear *@z3ntu*, I have build and tested new version whole day and have not faced any issues.
Also I fix a small formatting inconsistency (mix uppercaseand lowercase in hex-numbers) and switch to color output in make file.

**Comment by z3ntu** [comment]
Thanks a lot for the PR and the Huntsman Elite scroll wheel fix!

**Comment by beranat** [comment]
All thanks only to You - my push request fixs that I need, but your openrazer driver maintenance and support targets for all peoples good.

**Comment by Gotankssr** [comment]
Thanks for your work. I experience the same issue as Rincewind70, also on Linux mint. Apparently it does not recognize the keyboard layout and then falsely puts it to gb, whereas for me it should be us. Fixed this by putting the layout to 'kb-us' instead of 'kb-gb' if the keyboard layout does not exists in the profile_editor.js file. I'm new to github, so sorry if this is not the right way to let this know.
